opt: fix propagate_input_ordering implementation for CTEs

RaduBerinde opened the pull request:
Fixes #69288.

#### opt: check WithScan columns in CheckExpr

This change adds an assertion in CheckExpr that verifies that the
WithScan columns exist in the binding. The new assertion fires for
CTEs with `propagate_input_ordering` and for recursive CTEs. These
bugs are addressed in subsequent commits.

Release note: None

#### opt: fix binding for recursive CTE

The recursive CTE uses a dummy expression as the binding. The output
columns of this expression were not correct - they were the columns
from the initial expression, not the special "recursive" columns.

There is no known real-world consequence of this bug.

Release note: None

#### op: rename testfiles for propagate_input_ordering

Release note: None

#### opt: rename opttester flag for preserve-input-ordering

Release note: None

#### opt: keep track of PropagateInputOrdering in memo

This setting affects plans, so we need to check it when reusing memos.

Release note: None

#### opt: fix propagate_input_ordering implementation for CTEs

The current implementation of `propagate_input_ordering` breaks when
multiple CTEs are used: it causes the WithScan operators to contain
invalid columns. This is not a simple implementation bug - I just
don't see how this can work with multiple CTEs.

This change removes this logic and replaces it with a different idea -
we now optionally store a `BindingOrdering` in the `With` operator,
which controls the order in which the rows are materialized. Along with
forcing materialization when the `propagate_input_ordering` flag is
used, this guarantees that rows will be processed from that CTE in
order (all materialization happens on the gateway).

Release note: None

RaduBerinde committed:
opt: fix propagate_input_ordering implementation for CTEs

The current implementation of `propagate_input_ordering` breaks when
multiple CTEs are used: it causes the WithScan operators to contain
invalid columns. This is not a simple implementation bug - I just
don't see how this can work with multiple CTEs.

This change removes this logic and replaces it with a different idea -
we now optionally store a `BindingOrdering` in the `With` operator,
which control the order in which the rows are materialized. Along with
forcing materialization when the `propagate_input_ordering` flag is
used, this guarantees that rows will be processed from that CTE in
order (all materialization happens on the gateway).

Release note: None

RaduBerinde committed:
opt: keep track of PropagateInputOrdering in memo

This setting affects plans, so we need to check it when reusing memos.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/71380)
<!-- Reviewable:end -->

rytaft reviewed (commented):
Very nice!

> Reviewed 1 of 1 files at r1, 4 of 4 files at r2, 2 of 2 files at r3, 2 of 2 files at r4, 4 of 4 files at r5, 10 of 10 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-:-MllFC2T-w7CJhbszpHU:b-j7qk9t)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @RaduBerinde)

___
*[pkg/sql/opt/norm/ordering_funcs.go, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllCt9l9zUu52hxwdAN:-MllCt9l9zUu52hxwdAO:b2q6so9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/norm/ordering_funcs.go#L152)):*
> ```Go
> 
> // SimplifyWithBindingOrdering makes the ordering required by the WithBinding operator
> // less restrictive by removing optional columns, adding equivalent columns, and
> ```

removing optional columns -> adding optional columns?
___
*[pkg/sql/opt/norm/rules/ordering.opt, line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllDCKB99Rccr83oD0s:-MllDCKB99Rccr83oD0t:b-3h7iou) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/norm/rules/ordering.opt#L96)):*
> ```
> (Explain $input (SimplifyExplainOrdering $input $explainPrivate))
> 
> # TODO SimplifyWithBindingOrdering
> ```

What is the TODO for?
___
*[pkg/sql/opt/ops/relational.opt, line 1172 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllDx3D86woMtEqS1Vo:-MllDx3D86woMtEqS1Vp:b-coqv9r) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/ops/relational.opt#L1172)):*
> ```
>     # Ordering specifies the order required of the binding, and thus controls
>     # the order in which the rows are materialized. Can only be used when Mtr is
>     # set to always materialize.  Only used for the propagate_input_ordering
> ```

nit: remove extra space after period
___
*[pkg/sql/opt/optbuilder/testdata/propagate-input-ordering, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllEiZi9lod4H0lz5f1:-MllEiZi9lod4H0lz5f2:b2tndnw) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/optbuilder/testdata/propagate-input-ordering#L152)):*
> ```
> 
> # Do not preserve the subquery ordering because the parent scope has its own
> # ordering.
> ```

Does this comment need an update?
___
*[pkg/sql/opt/optbuilder/testdata/propagate-input-ordering, line 182 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllEnnt6KyCvmIWXBNg:-MllEnnvEG23Cf4Qjgt9:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/optbuilder/testdata/propagate-input-ordering#L182)):*
> ```
> 
> # Do not preserve the subquery ordering because the parent scope has its own
> # ordering.
> ```

ditto


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71380)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @mgartner and @rytaft)

___
*[pkg/sql/opt/norm/ordering_funcs.go, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllCt9l9zUu52hxwdAN:-MllRsQL7ctYbIfwKQpV:b-2vuolb) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/norm/ordering_funcs.go#L152)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

removing optional columns -> adding optional columns?
</blockquote></details>

I just copied this from above, I guess it means removing columns that can be optional from the ordering
___
*[pkg/sql/opt/norm/rules/ordering.opt, line 96 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllDCKB99Rccr83oD0s:-MllS4A-3iylNSR0qGHd:bfbpzdj) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/norm/rules/ordering.opt#L96)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

What is the TODO for?
</blockquote></details>

Fixed (leftover)
___
*[pkg/sql/opt/optbuilder/testdata/propagate-input-ordering, line 152 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllEiZi9lod4H0lz5f1:-MllSO4R2iR1DV1xBt8H:b-2l24x9) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/optbuilder/testdata/propagate-input-ordering#L152)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

Does this comment need an update?
</blockquote></details>

Good point, I tweaked the test
___
*[pkg/sql/opt/optbuilder/testdata/propagate-input-ordering, line 182 at r6](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-MllEnnt6KyCvmIWXBNg:-MllT8zL3YjJUKF2ZwlO:b6a0g36) ([raw file](https://github.com/cockroachdb/cockroach/blob/7b7c968073fb816ffa2be2e8664b5dcc97100119/pkg/sql/opt/optbuilder/testdata/propagate-input-ordering#L182)):*
<details><summary><i>Previously, rytaft (Rebecca Taft) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

Removed, looked like a duplicate test.


<!-- Sent from Reviewable.io -->

rytaft reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 3 of 3 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/71380#-:-MllUfol7lCNV7v1CYGK:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)



<!-- Sent from Reviewable.io -->

RaduBerinde committed:
opt: fix binding for recursive CTE

The recursive CTE uses a dummy expression as the binding. The output
columns of this expression were not correct - they were the columns
from the initial expression, not the special "recursive" columns.

There is no known real-world consequence of this bug.

Release note: None

RaduBerinde committed:
opt: check WithScan columns in CheckExpr

This change adds an assertion in CheckExpr that verifies that the
WithScan columns exist in the binding. The new assertion fires for
CTEs with `propagate_input_ordering` and for recursive CTEs. These
bugs are addressed in subsequent commits.

Release note: None

RaduBerinde committed:
opt: rename opttester flag for preserve-input-ordering

Release note: None

RaduBerinde committed:
op: rename testfiles for propagate_input_ordering

Release note: None
